interface hooks: prepare plug slot hooks (step 1)

stolowski opened the pull request:
First step towards hooks which collect plug/slot attributes at runtime. This addresses the comments of https://github.com/snapcore/snapd/pull/1767.

zyga commented on daemon/api_test.go:
```diff
@@ -2788,6 +2788,8 @@ func (s *apiSuite) TestInterfaces(c *check.C) {
 
 func (s *apiSuite) TestConnectPlugSuccess(c *check.C) {
 	d := s.daemon(c)
+	command := testutil.MockCommand(c, "snap", "")
```
Why are we mocking snap? Is it because of `snap run`? Are there any hooks being called here?

zyga commented on overlord/ifacestate/hook_handlers.go:
```diff
@@ -0,0 +1,56 @@
+	return nil
+}
+
+// SetupHooks sets hooks of InterfaceManager up
```
`setupHooks`

zyga commented on overlord/ifacestate/ifacemgr.go:
```diff
@@ -42,7 +43,11 @@ type InterfaceManager struct {
 // Extra interfaces can be provided for testing.
-func Manager(s *state.State, extra []interfaces.Interface) (*InterfaceManager, error) {
+func Manager(s *state.State, hookManager *hookstate.HookManager, extra []interfaces.Interface) (*InterfaceManager, error) {
+	if hookManager != nil {
```
If this is just for easier testing can you please add a comment saying that `// NOTE: hookManager is nil only when testing.`

zyga commented on overlord/ifacestate/ifacemgr.go:
```diff
@@ -72,12 +78,31 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
 	// different parameters so we cannot store the actual connection details).
-	summary := fmt.Sprintf(i18n.G("Connect %s:%s to %s:%s"),
+	plugHookSetup := &hookstate.HookSetup{
+		Snap:     plugSnap,
```
Don't you need the snap revision here a well?

niemeyer replied:
Unless some sort of useful evidence shows up, we should probably kill the idea of running hooks of a given revision. So far we haven't made any use of it at all, so it's been just added complexity.

zyga commented on overlord/ifacestate/ifacemgr.go:
```diff
@@ -72,12 +78,31 @@ func Connect(s *state.State, plugSnap, plugName, slotSnap, slotName string) (*st
+	collectPlugAttr := hookstate.HookTask(s, summary, plugHookSetup, nil)
+
+	slotHookSetup := &hookstate.HookSetup{
+		Snap:     slotSnap,
```
Same question about snap revision

zyga commented on overlord/ifacestate/ifacemgr_test.go:
```diff
@@ -49,11 +51,12 @@ type interfaceManagerSuite struct {
 	restoreBackends func()
-
-	storeSigning *assertstest.StoreStack
+	command         *testutil.MockCmd
```
I'd just call this mockSnapCmd and always do it (if that's what you need) so that there are less tests and it is more obvious what's going on. Feel free to ignore though as I'm just nitpicking here based on personal preference.

zyga commented on overlord/ifacestate/ifacemgr_test.go:
```diff
@@ -93,13 +101,31 @@ func (s *interfaceManagerSuite) TearDownTest(c *C) {
+}
+
+func (s *interfaceManagerSuite) settle(c *C) {
+	for i := 0; i < 50; i++ {
```
I really dislike this settle thing. We should have predictable and deterministic behaviour on how many times we need to call it and build some safeguards that check that after X iterations we're still changing the state so we're giving up.

Still, don't block on this. It's a general comment.

niemeyer replied:
I think we should actually fix this as it doesn't sound reasonable to just run 50 times blindly. We're not doing that elsewhere, so there should be a good way to handle this. Let's please discuss this tomorrow.

zyga commented on overlord/ifacestate/ifacemgr_test.go:
```diff
@@ -113,7 +139,14 @@ func (s *interfaceManagerSuite) TestConnectTask(c *C) {
-	task := ts.Tasks()[0]
+	i := 0
+	task := ts.Tasks()[i]
+	c.Check(task.Kind(), Equals, "run-hook")
```
Can you please check that each of those tasks carries the right meta-data about what hook to run?

niemeyer commented on snap/hooktypes.go:
```diff
@@ -26,6 +26,7 @@ import (
 var supportedHooks = []*HookType{
 	newHookType(regexp.MustCompile("^prepare-device$")),
 	newHookType(regexp.MustCompile("^configure$")),
+	newHookType(regexp.MustCompile("^prepare-(?:plug|slot)-[a-zA-Z0-9_\\-]+$")),
```
Same about chars constraint.

stolowski replied:
I agree about the settle thing in the test being ugly, but the in fact it is used in a couple of places already; I'm happy to discuss this of course, but perhaps it should be addressed in a separate PR.

~/gocode/src/github.com/snapcore/snapd/overlord $ grep -r -C1 "i < 50" *
assertstate/assertmgr_test.go-  // XXX: would like to use Overlord.Settle but not enough control there
assertstate/assertmgr_test.go:  for i := 0; i < 50; i++ {
assertstate/assertmgr_test.go-          s.mgr.Ensure()
--
devicestate/devicemgr_test.go-func (s *deviceMgrSuite) settle() {
devicestate/devicemgr_test.go:  for i := 0; i < 50; i++ {
devicestate/devicemgr_test.go-          s.hookMgr.Ensure()
--
snapstate/snapmgr_test.go-      // FIXME: use the real settle here
snapstate/snapmgr_test.go:      for i := 0; i < 50; i++ {
snapstate/snapmgr_test.go-              s.snapmgr.Ensure()
--
snapstate/booted_test.go-func (bs *bootedSuite) settle() {
snapstate/booted_test.go:       for i := 0; i < 50; i++ {
snapstate/booted_test.go-               bs.snapmgr.Ensure()

stolowski commented:
This branch is ready for a re-review; I've addressed all the comments so far, except for the problem of the loop over settle in the test (which as discussed is a problem affecting other tests too and needs a proper fix which doesn't belong to this PR). The remark about a nice-to-have test will be addressed in one of the followup PRs when the entire feature becomes more complete.
